File: algorithms/sorting/tim-sort/go/tim_sort.go
package timsort

const RUN = 32

func TimSort(arr []int) {
	n := len(arr)
	for i := 0; i < n; i += RUN {
		insertionSort(arr, i, min((i+RUN-1), (n-1)))
	}

	for size := RUN; size < n; size = 2 * size {
		for left := 0; left < n; left += 2 * size {
			mid := left + size - 1
			right := min((left + 2*size - 1), (n - 1))

			if mid < right {
				merge(arr, left, mid, right)
			}
		}
	}
}

func insertionSort(arr []int, left, right int) {
	for i := left + 1; i <= right; i++ {
		temp := arr[i]
		j := i - 1
		for j >= left && arr[j] > temp {
			arr[j+1] = arr[j]
			j--
		}
		arr[j+1] = temp
	}
}

func merge(arr []int, l, m, r int) {
	len1 := m - l + 1
	len2 := r - m
	left := make([]int, len1)
	right := make([]int, len2)

	for i := 0; i < len1; i++ {
		left[i] = arr[l+i]
	}
	for i := 0; i < len2; i++ {
		right[i] = arr[m+1+i]
	}

	i, j, k := 0, 0, l

	for i < len1 && j < len2 {
		if left[i] <= right[j] {
			arr[k] = left[i]
			i++
		} else {
			arr[k] = right[j]
			j++
		}
		k++
	}

	for i < len1 {
		arr[k] = left[i]
		k++
		i++
	}

	for j < len2 {
		arr[k] = right[j]
		k++
		j++
	}
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

File: algorithms/sorting/tim-sort/c/tim_sort.c
#include "tim_sort.h"

#define MIN(a,b) (((a)<(b))?(a):(b))

const int RUN = 32;

static void insertion_sort(int arr[], int left, int right) {
    for (int i = left + 1; i <= right; i++) {
        int temp = arr[i];
        int j = i - 1;
        while (j >= left && arr[j] > temp) {
            arr[j + 1] = arr[j];
            j--;
        }
        arr[j + 1] = temp;
    }
}

static void merge(int arr[], int l, int m, int r) {
    int len1 = m - l + 1, len2 = r - m;
    int left[len1], right[len2];
    
    for (int i = 0; i < len1; i++)
        left[i] = arr[l + i];
    for (int i = 0; i < len2; i++)
        right[i] = arr[m + 1 + i];
        
    int i = 0, j = 0, k = l;
    
    while (i < len1 && j < len2) {
        if (left[i] <= right[j]) {
            arr[k] = left[i];
            i++;
        } else {
            arr[k] = right[j];
            j++;
        }
        k++;
    }
    
    while (i < len1) {
        arr[k] = left[i];
        k++;
        i++;
    }
    
    while (j < len2) {
        arr[k] = right[j];
        k++;
        j++;
    }
}

void tim_sort(int arr[], int n) {
    for (int i = 0; i < n; i += RUN)
        insertion_sort(arr, i, MIN((i + RUN - 1), (n - 1)));
        
    for (int size = RUN; size < n; size = 2 * size) {
        for (int left = 0; left < n; left += 2 * size) {
            int mid = left + size - 1;
            int right = MIN((left + 2 * size - 1), (n - 1));
            
            if (mid < right)
                merge(arr, left, mid, right);
        }
    }
}
